File: packages/mylife-money/src/business/bots.ts
import cron from 'node-cron';
import { getStoreCollection, notifyView, createLogger, getService } from 'mylife-tools-server';

const logger = createLogger('mylife:money:business:bots');

export function createBot(values) {
  validateSchedule(values);
  const bots = getStoreCollection('bots');
  return bots.set(bots.entity.newObject(values));
}

export function updateBot(values) {
  validateSchedule(values);
  const bots = getStoreCollection('bots');
  let existingBot = bots.get(values._id);
  existingBot = bots.entity.setValues(existingBot, values);
  existingBot = bots.set(existingBot);
  return existingBot;
}

export function clearBotState(id: string) {
  logger.debug(`clea bot state '${id}'`);
  const bots = getStoreCollection('bots');
  let bot = bots.get(id);
  bot = bots.entity.getField('state').resetValue(bot);
  bots.set(bot);
}

function validateSchedule(values) {
  // note: cannot easily validate isomorophic
  if (values.schedule && !cron.validate(values.schedule)) {
    throw new Error(`Invalid schedule '${values.schedule}'`);
  }
}

export function deleteBot(id: string) {
  logger.debug(`drop bot '${id}'`);
  const bots = getStoreCollection('bots');
  const botRuns = getStoreCollection('bot-runs');

  const runs = botRuns.filter(run => run.bot === id);
  logger.debug(`found children runs: ${JSON.stringify(runs.map(run => run._id))}`);

  for (const run of runs) {
    botRuns.delete(run._id);
  }

  bots.delete(id);
}

export function notifyBots(session) {
  const bots = getStoreCollection('bots');
  return notifyView(session, bots.createView());
}

export function startBot(botId: string) {
  const service = getService('bot-service');
  service.startBot(botId);
}

export function startBotRun(botId: string): string {
  const botRuns = getStoreCollection('bot-runs');
  const values = {
    bot: botId,
    start: new Date(),
    logs: [],
  };

  const run = botRuns.set(botRuns.entity.newObject(values));
  return run._id;
}

export function endBotRun(runId: string, result: 'success' | 'warning' | 'error') {
  const botRuns = getStoreCollection('bot-runs');
  const run = botRuns.get(runId);
  const values = {
    end: new Date(),
    result
  };

  botRuns.set(botRuns.entity.setValues(run, values));
}

export function logBotRun(runId: string, severity: 'debug' |'info' |'warning' |'error' |'fatal', message: string) {
  const botRuns = getStoreCollection('bot-runs');
  const run = botRuns.get(runId);
  const log = { date: new Date(), severity, message };
  const values = { logs: [...run.logs, log] };

  botRuns.set(botRuns.entity.setValues(run, values));
}

export function updateBotState(botId: string, state: any) {
  const values = { _id: botId, state };
  updateBot(values);
}

export function notifyBotRuns(session, botId: string) {
  const botRuns = getStoreCollection('bot-runs');
  return notifyView(session, botRuns.createView(run => run.bot === botId));
}

File: go/mylife-money/pkg/business/bots.go
package business

import (
	"fmt"
	"mylife-money/pkg/entities"
	"mylife-tools-server/services/notification"
	"mylife-tools-server/services/sessions"
	"mylife-tools-server/services/store"
)

func NotifyBots(session *sessions.Session) uint64 {
	view := store.GetMaterializedView[*entities.Bot]("bots")
	viewId := notification.NotifyView(session, view)
	return viewId
}

var botStartHandler func(entities.BotType) error

func SetBotStartHandler(handler func(entities.BotType) error) {
	botStartHandler = handler
}

func ResetBotStartHandler() {
	botStartHandler = nil
}

func StartBot(id string) error {
	// Note: entities.BotType(typ) == string(id) for bots, cf bots view
	typ := entities.BotType(id)

	handler := botStartHandler
	if handler == nil {
		return fmt.Errorf("Bot start handler not set")
	}

	return handler(typ)
}

func GetAccountId(code string) (string, error) {
	accounts := store.GetCollection[*entities.Account]("accounts")

	list := accounts.Filter(func(a *entities.Account) bool {
		return a.Code() == code
	})

	if len(list) == 0 {
		return "", fmt.Errorf("account %s not found", code)
	}

	if len(list) > 1 {
		return "", fmt.Errorf("multiple accounts found for code %s", code)
	}

	return list[0].Id(), nil
}
